django_components/component.py

```diff
@@ -60,9 +62,93 @@ def __new__(mcs, name, bases, attrs):
+    """
+    Check if component's HTML, JS and CSS files refer to files in the same directory
+    as the component class. If so, modify the attributes so the class Django's rendering
+    will pick up these files correctly.
+    """
+
+    # Prepare all possible directories we need to check when searching for
+    # component's template and media files
+    candidate_dirs = settings.STATICFILES_DIRS or []
+    candidate_dirs.extend(
+        str(dir_path) for template_backend in settings.TEMPLATES or [] for dir_path in template_backend.get("DIRS", [])
+    )
+
+    # Get the directory where the component class is defined
+    try:
+        abs_comp_dir, comp_dir = _get_dir_path_from_component_module_path(attrs["__module__"], candidate_dirs)
+    except RuntimeError:
+        # If no dir was found, we assume that the path is NOT relative to the component dir
+        return
+
+    # Check if filepath refers to a file that's in the same directory as the component class.
+    def resolve_file(filepath: str):
+        maybe_resolved_filepath = os.path.join(abs_comp_dir, filepath)
+        component_import_filepath = os.path.join(comp_dir, filepath)
+
+        if os.path.isfile(maybe_resolved_filepath):
+            return component_import_filepath
+        return filepath
+
+    # Check if template name is a local file or not
+    if "template_name" in attrs and attrs["template_name"]:
+        attrs["template_name"] = resolve_file(attrs["template_name"])
+
+    if "Media" in attrs:
+        media = attrs["Media"]
+
+        # Now check the same for CSS files
+        if hasattr(media, "css") and isinstance(media.css, dict):
+            for media_type, path_list in media.css.items():
+                media.css[media_type] = [resolve_file(filepath) for filepath in path_list]
+
+        # And JS
+        if hasattr(media, "js") and isinstance(media.js, list):
+            media.js = [resolve_file(filepath) for filepath in media.js]
+
+
+def _get_dir_path_from_component_module_path(component_module_path: str, candidate_dirs: list[str]):
+    # Transform python module notation "pkg.module.name" to file path "pkg/module/name"
+    comp_path = os.sep.join(component_module_path.split("."))
+    comp_dir_rel_path = os.path.dirname(comp_path)
+
+    # Django resolves the template's relative to the values in TEMPLATE.DIRS
+    # so we have to adjust the component dir path to match one of the template dirs.
+    cwd = os.getcwd()
+    comp_dir_abs_path = os.path.join(cwd, comp_dir_rel_path)
+
+    # From all dirs defined in TEMPLATES.DIRS and STATICFILES_DIRS, find one that
+    # holds the component file.
+    root_dir_abs = None
+    for candidate_dir in candidate_dirs:
+        candidate_dir_abs = os.path.abspath(candidate_dir)
+        if comp_dir_abs_path.startswith(candidate_dir_abs):
+            # Get the common part of the path
+            root_dir_abs = comp_dir_abs_path[: len(candidate_dir_abs)]
+            break
+
+    if root_dir_abs is None:
```

Comment:
Good catch. I think it would make sense to log an error at the debug level, so that enabling debugging could allow people to see what paths are resolved when looking for the correct file. We get a lot of questions from people that are unsure of how the configure paths correctly.

Reply:
I just got around to look into logging. I wanted to set the log level via settings, but not sure how to proceed. How are settings handled internally in this project?

From the README, I would assume that it's configured via the `COMPONENTS` Django settings field. There are a few places where it's used like so:
```py
getattr(settings, "COMPONENTS", {}).get("RENDER_DEPENDENCIES", False)
```

However, then there's also the `AppSettings`, which is used in more places, like so:
```py
app_settings.CONTEXT_BEHAVIOR == "isolated"
```

So I would probably use `AppSettings` to define and access the log level settings.

Reply:
Ok turns out that I didn't need to touch the AppSettings, as logging can be configured [via Django](https://docs.djangoproject.com/en/5.0/topics/logging/).

I've added:
- Logging example to `sampleproject`
- Section on Logging to README
- A few logging statements

<details>
<summary>
Here you can see that the logs are printed only if log level is set to DEBUG for `django_components`
</summary>

<img width="429" alt="Screenshot 2024-03-17 at 20 00 27" src="https://github.com/EmilStenstrom/django-components/assets/25986782/2531e268-27d8-45e2-b9ac-879a7c19d48f">

<img width="676" alt="Screenshot 2024-03-17 at 20 00 46" src="https://github.com/EmilStenstrom/django-components/assets/25986782/d75d1d01-97bc-43c2-bdf2-615b1102b5e7">

</details>


<details>
<summary>
And here are some examples of what's logged when component uses files relative to the component vs not
</summary>

<img width="728" alt="Screenshot 2024-03-17 at 20 10 11" src="https://github.com/EmilStenstrom/django-components/assets/25986782/89721588-2321-47c6-849c-1e673db56d70">

<img width="732" alt="Screenshot 2024-03-17 at 20 10 21" src="https://github.com/EmilStenstrom/django-components/assets/25986782/4af4a10b-c8c0-4bec-a01f-b981a0a0968c">

</details>